### PR TITLE
[CLOUD-78] authz: re-enable user-centric perms syncing for user code host conns

### PR DESCRIPTION
**Pull request by unknwon**
_This is a reworked implementation of #25884 since [the original implementation was reverted](https://github.com/sourcegraph/sourcegraph/pull/26636) due to [#INC-32: CloudKMS rate limit exceeded](https://app.incident.io/incidents/32)._

For small orgs EA, we need to (re-)enable user-centric permissions syncing based on user code host connections because login connections does not have `repo` scope granted (which is required to access GitHub repo APIs).

This implies that:
1. Users need to connect personal code host connection in order to see private repositories added to their organizations on Sourcegraph.
2. When users go through OAuth flow, they need to grant access from the GitHub organization they're intended to add repositories to Sourcegraph.

This PR adds back the code that is functionally equivalent to https://github.com/sourcegraph/sourcegraph/pull/23018.

### What's changed compare to #25884

Instead of listing all code host connections that includes over 3k every 5 seconds, we only list site and org-level code host connections, which for EA should only be around 20.

Then, for user-centric permissions syncing, we expect around "number of user per minute" KMS API calls, e.g. 100 users being members of organization has connected code host connection, 100 KMS API calls per minute.

### Follow up

This reworked implementation is only good enough for EA, and how to solve/reduce KMS API usage is beyond the purpose of this PR and should be addressed as an independent task with a holistic solution. I will create Jira ticket for it.

---

Jira: CLOUD-78

### CODE REVIEW

File: internal/database/external_services.go [hunk 1]
```diff
@@ -145,8 +145,12 @@ type ExternalServicesListOptions struct {
 	// When specified, only include external services with the given IDs.
 	IDs []int64
 	// When true, only include external services not under any namespace (i.e. owned
-	// by all site admins), and value of NamespaceUserID is ignored.
+	// by all site admins), and values of NamespaceUserID, NamespaceOrgID and
+	// ExcludeNamespaceUser are ignored.
 	NoNamespace bool
+	// When true, will exclude external services under any user namespace, and
+	// values of NamespaceUserID and NamespaceOrgID are ignored.
+	ExcludeNamespaceUser bool
```
[the remark this answers is not in the document]
Comment (unknwon) [reply internal/database/external_services.go:153]: @kopancek Which lines you're referring that you saw "the behavior". I'm referring to the following there `NamespaceOrgID` is only consulted in an `else-if` branch (i.e. `ExcludeNamespaceUser` is `false`):

https://github.com/sourcegraph/sourcegraph/blob/08ee4f4d7890795e2119f3c2e53ae514104d010e/internal/database/external_services.go#L183-L188

File: enterprise/cmd/repo-updater/internal/authz/perms_syncer.go [hunk 2]
```diff
@@ -407,6 +408,142 @@ func (s *PermsSyncer) fetchUserPermsViaExternalAccounts(ctx context.Context, use
...
+	//  do the user-centric syncing.
+	if !has {
+		return []uint32{}, nil
+	}
+
+	svcs, err := database.ExternalServicesWith(s.reposStore).List(ctx,
+		database.ExternalServicesListOptions{
+			NamespaceUserID: userID,
+			Kinds:           []string{extsvc.KindGitHub, extsvc.KindGitLab},
+		},
```
[the remark this answers is not in the document]
Comment (unknwon) [reply enterprise/cmd/repo-updater/internal/authz/perms_syncer.go:446]: Chatted with @kopancek on a call last Friday, posting summary here for future readers. In short, org-level code host connections are only used for repo-centric permissions syncing, same as site-level code host connections.

File: internal/database/external_services.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kopancek) [reply internal/database/external_services.go:153]: Hmm, in that case I think we should change the logic in a follow up PR. What I would expect:
- `NoNamespace` -> any namespace is ignored
- `ExcludeNamespaceUser` -> only user namespace is ignored. If we supply `NamespaceOrgID`, it is still applied
- `NamespaceUserID`, `NamespaceOrgID` are mutually exclusive

[the remark this answers is not in the document]
Comment (unknwon) [reply internal/database/external_services.go:153]: Sounds good, PR sent: https://github.com/sourcegraph/sourcegraph/pull/27004

**Pull request merged**